server/src/main/java/org/elasticsearch/action/admin/indices/datastream/CreateDataStreamAction.java

```diff
@@ -116,7 +135,37 @@ protected AcknowledgedResponse read(StreamInput in) throws IOException {
         @Override
         protected void masterOperation(Task task, Request request, ClusterState state,
                                        ActionListener<AcknowledgedResponse> listener) throws Exception {
-            listener.onResponse(new AcknowledgedResponse(true));
+            clusterService.submitStateUpdateTask("create-data-stream [" + request.name + "]",
+                new ClusterStateUpdateTask(Priority.HIGH) {
+
+                    @Override
+                    public TimeValue timeout() {
+                        return request.masterNodeTimeout();
+                    }
+
+                    @Override
+                    public void onFailure(String source, Exception e) {
+                        listener.onFailure(e);
+                    }
+
+                    @Override
+                    public ClusterState execute(ClusterState currentState) throws Exception {
+                        if (currentState.metaData().dataStreams().containsKey(request.name)) {
```

Comment:
Perhaps extract the body of the execute method in a static method, so that is easy to write unit tests?
(and maybe the same for the other transport actions?)